internal/database/users.go

```diff
@@ -675,6 +677,44 @@ func logUserDeletionEvents(ctx context.Context, db DB, ids []int32, name Securit
 	db.SecurityEventLogs().LogEventList(ctx, events)
 }
 
+// RecoverList recovers a list of users by their IDs.
+func (u *userStore) RecoverUsersList(ctx context.Context, ids []int32) ([]int32, error) {
+	tx, err := u.Transact(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = tx.Done(err) }()
+	userIDs := make([]*sqlf.Query, len(ids))
+	for i := range ids {
+		userIDs[i] = sqlf.Sprintf("%d", ids[i])
+	}
+	idsCond := sqlf.Join(userIDs, ",")
+
+	affectedUserID, err := basestore.ScanInt32s(tx.Query(ctx, sqlf.Sprintf("UPDATE users SET deleted_at=NULL, updated_at=now() WHERE id IN (%s) AND deleted_at IS NOT NULL RETURNING id", idsCond)))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tx.Exec(ctx, sqlf.Sprintf("INSERT INTO names(name, user_id) SELECT username, id FROM users WHERE id IN(%s)", idsCond)); err != nil {
+		return nil, err
+	}
+
+	//todo: @mucles - need add where condition to query to only update if deleted_at is within 10 secs of users.deleted_at
+	if err := tx.Exec(ctx, sqlf.Sprintf("UPDATE user_external_accounts SET deleted_at=NULL, updated_at=now() WHERE user_id IN (%s) AND deleted_at IS NOT NULL", idsCond)); err != nil {
+		return nil, err
+	}
+
+	if err := tx.Exec(ctx, sqlf.Sprintf("UPDATE org_invitations SET deleted_at=NULL WHERE deleted_at IS NOT NULL and revoked_at IS NOT NULL and expires_at >= now() AND (sender_user_id IN (%s) OR recipient_user_id IN (%s))", idsCond, idsCond)); err != nil {
```

Comment:
We actually want `revoked_at IS NULL` here, because that means the invitation is NOT revoked.
```suggestion
	if err := tx.Exec(ctx, sqlf.Sprintf("UPDATE org_invitations SET deleted_at=NULL WHERE deleted_at IS NOT NULL and revoked_at IS NULL and expires_at >= now() AND (sender_user_id IN (%s) OR recipient_user_id IN (%s))", idsCond, idsCond)); err != nil {
```

Also, not sure it makes sense to revive org invitations at all, but I am not here to decide what the feature should accomplish.

Reply:
Good catch, thank you! It would make sense if the revived users could still access pending org invites, but I wonder if it is necessary since the org invite could just be resent again. If the org_invites aren't revived, we should also do hard delete org invites rather than : https://github.com/sourcegraph/sourcegraph/blob/main/internal/database/users.go#L552